// @remix-project/remixd/src/services/truffleClient.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TruffleClient = void 0;
const plugin_1 = require("@remixproject/plugin");
const { spawn } = require('child_process'); // eslint-disable-line
class TruffleClient extends plugin_1.PluginClient {
    constructor(readOnly = false) {
        super();
        this.readOnly = readOnly;
        this.methods = ['compile'];
    }
    setWebSocket(websocket) {
        this.websocket = websocket;
    }
    sharedFolder(currentSharedFolder) {
        this.currentSharedFolder = currentSharedFolder;
    }
    compile() {
        return new Promise((resolve, reject) => {
            if (this.readOnly) {
                const errMsg = '[Truffle Compilation]: Cannot compile in read-only mode';
                return reject(new Error(errMsg));
            }
            const cmd = `truffle compile`;
            const options = { cwd: this.currentSharedFolder, shell: true };
            const child = spawn(cmd, options);
            let result = '';
            let error = '';
            child.stdout.on('data', (data) => {
                const msg = `[Truffle Compilation]: ${data.toString()}`;
                console.log('\x1b[32m%s\x1b[0m', msg);
                result += msg + '\n';
            });
            child.stderr.on('data', (err) => {
                error += `[Truffle Compilation]: ${err.toString()} \n`;
            });
            child.on('close', () => {
                if (error && result)
                    resolve(error + result);
                else if (error)
                    reject(error);
                else
                    resolve(result);
            });
        });
    }
}
exports.TruffleClient = TruffleClient;
